diff --git a/doc/release/release_dev.rst b/doc/release/release_dev.rst
index 12a1793289..e0effa5146 100644
--- a/doc/release/release_dev.rst
+++ b/doc/release/release_dev.rst
@@ -27,6 +27,9 @@ Improvements
 API Changes
 -----------
 
+- ``colorconv.separate_stains`` and ``colorconv.combine_stains`` now uses
+  base10 instead of the natural logarithm as discussed in issue #2995.
+
 
 
 Bugfixes
diff --git a/skimage/color/colorconv.py b/skimage/color/colorconv.py
index 6cafc0abdc..6f56c4680b 100644
--- a/skimage/color/colorconv.py
+++ b/skimage/color/colorconv.py
@@ -1413,7 +1413,7 @@ def separate_stains(rgb, conv_matrix):
     """
     rgb = dtype.img_as_float(rgb, force_copy=True)
     rgb += 2
-    stains = np.reshape(-np.log(rgb), (-1, 3)) @ conv_matrix
+    stains = np.reshape(-np.log10(rgb), (-1, 3)) @ conv_matrix
     return np.reshape(stains, rgb.shape)
 
 
@@ -1474,7 +1474,7 @@ def combine_stains(stains, conv_matrix):
 
     stains = dtype.img_as_float(stains)
     logrgb2 = -np.reshape(stains, (-1, 3)) @ conv_matrix
-    rgb2 = np.exp(logrgb2)
+    rgb2 = np.power(10, logrgb2)
     return rescale_intensity(np.reshape(rgb2 - 2, stains.shape),
                              in_range=(-1, 1))
 
diff --git a/skimage/color/tests/test_colorconv.py b/skimage/color/tests/test_colorconv.py
index 767ab59809..0324869a3f 100644
--- a/skimage/color/tests/test_colorconv.py
+++ b/skimage/color/tests/test_colorconv.py
@@ -192,10 +192,11 @@ def test_hdx_rgb_roundtrip(self):
         img_rgb = self.img_rgb
         conv = combine_stains(separate_stains(img_rgb, hdx_from_rgb),
                               rgb_from_hdx)
-        assert_equal(img_as_ubyte(conv), img_rgb)
+        with expected_warnings('precision loss'):
+            assert_equal(img_as_ubyte(conv), img_rgb)
 
-    # RGB<->HDX roundtrip with ubyte image
-    def test_hdx_rgb_roundtrip(self):
+    # RGB<->HDX roundtrip with float image
+    def test_hdx_rgb_roundtrip_float(self):
         from skimage.color.colorconv import hdx_from_rgb, rgb_from_hdx
         img_rgb = img_as_float(self.img_rgb)
         conv = combine_stains(separate_stains(img_rgb, hdx_from_rgb),
