diff --git a/skimage/measure/tests/test_blur_effect.py b/skimage/measure/tests/test_blur_effect.py
index eaef61b87..8f3a016f6 100644
--- a/skimage/measure/tests/test_blur_effect.py
+++ b/skimage/measure/tests/test_blur_effect.py
@@ -4,9 +4,9 @@ from skimage.color import rgb2gray
 from skimage.data import astronaut, cells3d
 from skimage.filters import gaussian
 from skimage.measure import blur_effect
+from skimage._shared.testing import fetch

 image = astronaut()
-image_3d = cells3d()[:, 1, :, :]  # grab just the nuclei


 def test_blur_effect():
@@ -46,6 +46,10 @@ def test_blur_effect_channel_axis():

 def test_blur_effect_3d():
     """Test that the blur metric works on a 3D image."""
+    # Fetching the data through the testing module will
+    # cause the test to skip if pooch isn't installed.
+    fetch("data/cells3d.tif")
+    image_3d = cells3d()[:, 1, :, :]  # grab just the nuclei
     B0 = blur_effect(image_3d)
     B1 = blur_effect(gaussian(image_3d, sigma=1))
     B2 = blur_effect(gaussian(image_3d, sigma=4))
